cmd/snap-confine/ns-support.c

```diff
@@ -306,10 +308,172 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
+		// task or wait for anything.
+		if (prctl(PR_SET_PDEATHSIG, SIGINT, 0, 0, 0) < 0) {
+			die("cannot set parent process death notification signal to SIGINT");
+		}
+		// Check that parent process is still alive. If this is the case then
+		// we can *almost* reliably rely on the PR_SET_PDEATHSIG signal to wake
+		// us up from eventfd_read() below. In the rare case that the PID
+		// numbers overflow and the now-dead parent PID is recycled we will
+		// still hang forever on the read from eventfd below.
+		debug("ensuring that parent process is still alive");
+		if (kill(parent, 0) < 0) {
+			switch (errno) {
+			case ESRCH:
+				debug("parent process has already terminated");
+				abort();
+			default:
+				die("cannot ensure that parent process is still alive");
+				break;
+			}
+		}
+
+		debug("joining the namespace that we are about to probe");
+		// Move to the mount namespace of the snap we're trying to inspect.
+		if (setns(mnt_fd, CLONE_NEWNS) < 0) {
+			die("cannot join the mount namespace in order to inspect it");
+		}
+		// Check if the namespace needs to be discarded.
+		//
+		// TODO: enable this for core distributions. This is complex because on
+		// core the rootfs is mounted in initrd and is _not_ changed (no
+		// pivot_root) and the base snap is again mounted (2nd time) by
+		// systemd. This makes us end up in a situation where the outer base
+		// snap will never match the rootfs inside the mount namespace.
+		bool should_discard =
+		    is_classic ? should_discard_current_ns(base_snap_dev) :
+		    false;
+
+		// Send this back to the parent: 2 - discard, 1 - keep.
+		// Note that we cannot just use 0 and 1 because of the semantics of eventfd(2).
+		debug
+		    ("sending information about the state of the mount namespace (%s)",
+		     should_discard ? "discard" : "keep");
+		if (eventfd_write
+		    (event_fd,
+		     should_discard ? SC_DISCARD_YES : SC_DISCARD_NO) < 0) {
+			die("cannot send information about the state of the mount namespace");
+		}
+		// Exit, we're done.
+		debug
+		    ("support process for mount namespace inspection is about to finish");
+		exit(0);
+	}
+	// This is back in the parent process.
+	//
+	// Enable a sanity timeout in case the read blocks for unbound amount of
+	// time. This will ensure we will not hang around while holding the lock.
+	// Next, read the value written by the child process.
+	sc_enable_sanity_timeout();
+	eventfd_t value = 0;
+	debug("receiving information about the state of the mount namespace");
+	if (eventfd_read(event_fd, &value) < 0) {
+		die("cannot receive information about the state of the mount namespace");
+	}
+	sc_disable_sanity_timeout();
+
+	// Wait for the child process to exit and collect its exit status.
+	errno = 0;
+	int status;
+	if (waitpid(child, &status, 0) < 0) {
+		die("cannot wait for the support process for mount namespace inspection");
+	}
+	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
```

Comment:
Something that occurred to me now. Can't we  an exit code instead of eventfd?

Reply:
Hmm, perhaps though I just reused the eventfd pattern from other parts of snap-confine. The idea is that we just want to peek inside and get an answer. This can be encoded in an exit status but it could also be (somehow) more confusing to remap things like "I got killed by singal, I got killed by other-stuff" away from our boolean flag. Unless you strongly feel about it I'd like to keep eventfd for this.